### DISCUSSION TITLE
ESP32-WROOM-32E-N8R2 with 2 MB SPI RAM and esp32spiram-20220618-v1.19.1.bin not working

### DISCUSSION THREAD

**Discussion by sseuf** [micropython/micropython#10875] [ESP32]
Hi,

I have flashed a [ESP32-WROOM-32E-N8R2](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf) (ESP32-D0WDR2-V3 with 2 MB of SPI RAM) with the latest esp32/GENERIC_SPIRAM Firmware [v1.19.1 (2022-06-18) .bin](https://micropython.org/resources/firmware/esp32spiram-20220618-v1.19.1.bin).

Up on boot I get the following output:
```
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 153911750, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12788
load:0x40080400,len:4176
entry 0x40080680
E (651) psram: Not a valid or known package id: 7
Re-enable cpu cache.

abort() was called at PC 0x40082e57 on core 0

Backtrace:0x400964a9:0x3ffe3b30 0x40096a99:0x3ffe3b50 0x4009a355:0x3ffe3b70 0x40
082e57:0x3ffe3be0 0x400d3ec5:0x3ffe3c50 0x40081e08:0x3ffe3c70 0x40078f5c:0x3ffe3
c90 |<-CORRUPTED


ELF file SHA256: 46bca36b7d6020a6

Rebooting...
```
After that, the device restarts endlessly.

I have tracked the error message `E (651) psram: Not a valid or known package id: 7` down in the esp-idf in the file [esp_psram_impl_quad.c](https://github.com/espressif/esp-idf/blob/8a7f6af6256f4c954053e08e819f55b93ddc1ff1/components/esp_psram/esp32/esp_psram_impl_quad.c#L873)

But the if-statement for the ESP32-D0WDR2-V3 is present in line 873: 
```
} else if (pkg_ver == EFUSE_RD_CHIP_VER_PKG_ESP32D0WDR2V3){
```
And EFUSE_RD_CHIP_VER_PKG_ESP32D0WDR2V3 is defined as 7 in [efuse_reg.h](https://github.com/espressif/esp-idf/blob/9b8c558e63d95b491355a22e69f247834e899b47/components/soc/esp32/include/soc/efuse_reg.h#L111).

So I searched the commit where the if-statement was added and it is [ce1ee3d](https://github.com/espressif/esp-idf/commit/ce1ee3d8aebaca13f4db802001ecd89dd89140ab) from Nov 12, 2021. The name of the file `esp_psram_impl_quad.c` was `spiram_psram.c` at this time. The module `esp_psram_impl_quad.o` is not found in the map-file [esp32spiram-20220618-v1.19.1.map](https://micropython.org/resources/firmware/esp32spiram-20220618-v1.19.1.map) but `spiram_psram.o` is.

Is the current `esp32spiram-20220618-v1.19.1` built with a esp-idf from earlier than Nov 12, 2021 or am I missing something?

I don't know much about building micropython, but I want to use the SPI RAM of the ESP32-WROOM-32E-N8R2. Does someone has a suggestion?

**Comment by PBrunot** [comment]
Building more recet Micropython image is not that hard. I was able to get it running in 2 hours on Windows + WSL from scratch and little previous experience.
1. Build micropython mpy-cross as per these docs: https://docs.micropython.org/en/latest/develop/gettingstarted.html#compile-and-build-the-code
2. Follow ESP32 specifics instructions from  :https://github.com/micropython/micropython/blob/master/ports/esp32/README.md
I suggest you build against ESP-IDF v4.4  (5.x not supported)

3. To create a specific board image, from ports/esp32 folder
```
make BOARD=GENERIC_SPIRAM
```
4. Flash the firmware.bin onto your device (file under ports/esp32/build-<board> folder)
